### PR TITLE
ZK-4855: some ListProxy methods set _dirty directly instead of callin…

### PR DESCRIPTION
**Pull request by DevChu**
…g setDirty, missing updates

### CODE REVIEW

File: zkbind/src/org/zkoss/bind/proxy/FormProxyHandler.java [hunk 1]
```diff
@@ -165,7 +167,17 @@ public Object invoke(Object self, Method method, Method proceed, Object[] args)
 	private void addCallbacks() {
 		_node.setOnDataChangeCallback(new Callback<Object[]>() {
 			public void call(Object[] data) {
-				_binding.getBinder().notifyChange(data[0], (String) data[1]);
+				String zkFormProxyNotifiedKey = "$$zkFormProxyNotified$$" + data[0];
```
Comment (scribetw) [inline zkbind/src/org/zkoss/bind/proxy/FormProxyHandler.java:170]: Attribute name is "$$zkFormProxyNotified$$", use Set<Pair<Object, String>>

**Review by scribetw — CHANGES_REQUESTED** [review]

**Review by DevChu — COMMENTED** [review]
write a case for same exec but later change

File: zkbind/src/org/zkoss/bind/impl/BinderImpl.java [hunk 2]
```diff
@@ -459,6 +464,14 @@ private void doPropertyChange(Object base, String prop) {
 			return;
 		}
 
+		//ZK-4855
+		if (base instanceof FormProxyObject) {
+			Execution execution = Executions.getCurrent();
+			Set<Pair<Object, String>> zkProxyNotified = execution != null ? (Set<Pair<Object, String>>) execution.getAttribute(ZKFORMPROXYNOTIFIEDKEY) : null;
+			if (zkProxyNotified != null && zkProxyNotified.size() > 0)
```
Comment (scribetw) [inline zkbind/src/org/zkoss/bind/impl/BinderImpl.java:471]: only `zkProxyNotified != null`

File: zkbind/src/org/zkoss/bind/proxy/ListProxy.java [hunk 3]
```diff
@@ -47,18 +47,20 @@ public E get(int index) {
 	}
 
 	public E set(int index, E element) {
-		_dirty = true;
-		return ((List<E>) getCache()).set(index, element);
+		E prevElement = ((List<E>) getCache()).set(index, element);
```
Comment (scribetw) [inline zkbind/src/org/zkoss/bind/proxy/ListProxy.java:50]: createProxyObject(element)

**Review by scribetw — CHANGES_REQUESTED** [review]

File: zkbind/src/org/zkoss/bind/proxy/ListProxy.java [hunk 4]
```diff
@@ -47,18 +47,20 @@ public E get(int index) {
 	}
 
 	public E set(int index, E element) {
-		_dirty = true;
```
Comment (scribetw) [inline zkbind/src/org/zkoss/bind/proxy/ListProxy.java:50]: check addAll

**Review by scribetw — CHANGES_REQUESTED** [review]

**Review by plijyfes — APPROVED** [review]

**Review by scribetw — APPROVED** [review]

**Pull request merged**